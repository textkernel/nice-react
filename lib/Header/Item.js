import React from 'react';
import PropTypes from 'prop-types';
import props from '../props';

const Item = ({ children, className, context, label, level, href, target, width }) => {
    const classes = [];

    if (children && level) {
        classes.push('has-children');
    }
    if (context) {
        classes.push('highlight', context);
    }
    if (width) {
        classes.push('fixed-width', `w${ width }`);
    }

    return (<li className={ classes.join(' ') || null }>
        <a href={ href } target={ target } className={ className }>
            { label }
        </a>
        { 
            children
            ? <ul>
                { React.Children.map(children, child => (
                    React.cloneElement(child, { level: level + 1 })
                )) }
            </ul>
            : null
        }
    </li>);
};

Item.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.array
    ]),
    className: PropTypes.string,
    context: PropTypes.oneOf(props.headerMenuContexts),
    label: PropTypes.string,
    level: PropTypes.number,
    href: PropTypes.string,
    target: PropTypes.string,
    width: PropTypes.oneOf(props.headerMenuWidths)
};

Item.defaultProps = {
    children: null,
    className: null,
    context: null,
    label: '',
    level: 0,
    href: '#',
    target: null,
    width: null
};

export default Item;
