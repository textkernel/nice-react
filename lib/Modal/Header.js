import React, { Component } from 'react';
import PropTypes from 'prop-types';

export default class ModalHeader extends Component {
    static propTypes = {
        children: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number
        ]),
        onClose: PropTypes.func
    }

    static defaultProps = {
        children: null,
        onClose: null
    }

    renderCloseBtn() {
        const { onClose } = this.props;

        if (!onClose) {
            return null;
        }
        return (
            <a href="#" className="nice-modal-close" onClick={ e => {
                e.preventDefault();
                onClose();
            }}>
                <i className="icon fa-times" />
            </a>
        );
    }

    render() {
        const { children } = this.props;

        if (!children) {
            return null;
        }

        return (
            <div className="nice-modal-header">
                { this.renderCloseBtn() }
                <h4>
                    { children }
                </h4>
            </div>
        );
    }
}
