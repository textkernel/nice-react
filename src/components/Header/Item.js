import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import props from '../../props';


const Item = ({ 
    children,
    className,
    context,
    label,
    level,
    onClick,
    href,
    target,
    width,
    style
}) => (
    <li
        style={style}
        className={ classNames({
            'has-children': children && level,
            highlight: context,
            [context]: context,
            'fixed-width': width,
            [`w${ width }`]: width
        }) }
    >
        <a
            href={ href }
            target={ target }
            className={ className }
            onClick={ e => {
                if (!onClick) {
                    return true;
                }
                onClick(e);
                return false;
            }}
        >
            { label }
        </a>
        { 
            children
            ? <ul>
                { React.Children.map(children, child => (
                    React.cloneElement(child, { level: level + 1 })
                )) }
            </ul>
            : null
        }
    </li>
);

Item.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.array
    ]),
    className: PropTypes.string,
    context: PropTypes.oneOf(props.headerMenuContexts),
    label: PropTypes.string,
    level: PropTypes.number,
    onClick: PropTypes.func,
    href: PropTypes.string,
    target: PropTypes.string,
    width: PropTypes.oneOf(props.headerMenuWidths),
    style: PropTypes.object,
};

Item.defaultProps = {
    children: null,
    className: null,
    context: null,
    label: '',
    level: 0,
    onClick: null,
    href: '#',
    target: null,
    width: null,
    style: {}
};

export default Item;
