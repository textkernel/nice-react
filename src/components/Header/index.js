export { default } from './Header';
export { default as HeaderMenu } from './Menu';
export { default as HeaderMenuItem } from './Item';
