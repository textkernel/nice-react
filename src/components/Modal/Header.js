import React, { Component } from 'react';
import classnames from 'classnames';
import PropTypes from 'prop-types';

class Header extends Component {
    renderCloseBtn() {
        const { onClose } = this.props;

        if (!onClose) {
            return null;
        }
        return (
            <a href="#" className="nice-modal-close" onClick={ e => {
                e.preventDefault();
                onClose();
            }}>
                <i className="icon fa-times" />
            </a>
        );
    }

    render() {
        const { children, className, style } = this.props;

        if (!children) {
            return null;
        }

        return (
            <div className={classnames('nice-modal-header', className)} style={style}>
                { this.renderCloseBtn() }
                <h4>
                    { children }
                </h4>
            </div>
        );
    }
}

Header.propTypes = {
    className: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number
    ]),
    style: PropTypes.object,
    onClose: PropTypes.func
};

Header.defaultProps = {
    className: null,
    children: null,
    style: {},
    onClose: null
};

export default Header;
