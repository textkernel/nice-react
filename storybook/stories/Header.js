import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, number, text } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';

import { Header } from 'nice-react';

const stories = storiesOf('Header', module);
stories.addDecorator((story, context) => withInfo('')(story)(context));
stories.addDecorator(withKnobs);

stories.add('Custom properties', () => {
    const product = text('Product', 'Nice! React Components');
    const svg = text('SVG logo (logo.svg)', '//nice.textkernel.nl/assets/textkernel.svg');
    const img = text('Logo (logo.img)', '//nice.textkernel.nl/assets/textkernel@2x.png');
    const width = number('Logo width (logo.width)', 172);
    const height = number('Logo height (logo.height)', 30);
    const title = text('Logo title (logo.title)', '');
    const href = text('Logo URL (logo.href)', '/');
    const fixed = boolean('Fixed header', false);
    const fluid = boolean('Fluid container layout', false);
    const placeholder = text('Input placeholder (form.placeholder)', 'Search...');
    const responsive = boolean('Responsive menu', true);
    const collapsed = !boolean('Menu is expanded', false);

    return (<Header logo={{
            svg,
            img,
            href,
            width,
            height,
            title
        }}
        form={{
            placeholder,
            onSubmit: value => {
                window.alert(value);
                return false;
            }
        }}
        collapsed={ collapsed }
        responsive={ responsive }
        fixed={ fixed }
        fluid={ fluid }
        product={ product }
    >
        <Header.Menu>
            <Header.Menu.Item label="Menu 1" context="primary" width="100">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
                <Header.Menu.Item label="Item 3">
                    <Header.Menu.Item label="Item 3.1" />
                    <Header.Menu.Item label="Item 3.2" />
                    <Header.Menu.Item label="Item 3.3" />
                    <Header.Menu.Item label="Item 3.4" />
                </Header.Menu.Item>
                <Header.Menu.Item label="Item 4" />
            </Header.Menu.Item>
            <Header.Menu.Item label="Menu 2" context="good" width="100">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
                <Header.Menu.Item label="Item 3">
                    <Header.Menu.Item label="Item 3.1" />
                    <Header.Menu.Item label="Item 3.2" />
                    <Header.Menu.Item label="Item 3.3" />
                    <Header.Menu.Item label="Item 3.4" />
                </Header.Menu.Item>
                <Header.Menu.Item label="Item 4" />
            </Header.Menu.Item>
            <Header.Menu.Item label="Textkernel" href="https://textkernel.com" target="_blank" context="brand" width="125" className="product-icon textkernel" />
        </Header.Menu>
        <Header.Menu alternate>
            <Header.Menu.Item label="Alternate 1">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
            </Header.Menu.Item>
            <Header.Menu.Item label="Alternate 2">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
            </Header.Menu.Item>
        </Header.Menu>
    </Header>);
});
