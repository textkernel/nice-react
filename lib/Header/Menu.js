import React from 'react';
import PropTypes from 'prop-types';
import Item from './Item';
import classNames from 'classnames';

const Menu = ({ children, alternate, level }) => (
    <ul className={ classNames('nice-header-menu', {
        'menu-alt': alternate
    }) }>
        { React.Children.map(children, child => (
            React.cloneElement(child, { level })
        )) }
    </ul>
);

Menu.propTypes = {
    alternate: PropTypes.bool,
    level: PropTypes.number
};

Menu.defaultProps = {
    alternate: false,
    level: 0
};

Menu.Item = Item;

export default Menu;
