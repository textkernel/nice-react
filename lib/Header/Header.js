import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Menu from './Menu';

class Header extends Component {
    renderImage() {
        const { svg, img, width = 170, height = 30, title } = this.props.logo || {};

        if (svg) {
            return (<svg width={ width } height={ height }>
                <image
                    xlinkHref={ svg }
                    src={ img }
                    width={ width }
                    height={ height } />
            </svg>);
        }
        return (<img
            src={ img }
            width={ width }
            height={ height }
            alt={ title }
        />);
    }

    renderBrand() {
        if (!this.props.logo) {
            return null;
        }

        const { href, title } = this.props.logo;

        return (<h1 className="nice-header-brand">
            <a href={ href } title={ title }>
                {
                    this.renderImage()
                }
            </a>
        </h1>);
    }

    renderForm() {
        const { form } = this.props;

        if (!form) {
            return null;
        }

        return (<form
            className="nice-header-form"
            onSubmit={ (e) => {
                e.preventDefault();

                if (!form.onSubmit) {
                    return false;
                }

                const value = e.target.getElementsByTagName('input')[0].value || null;

                if (!value) {
                    return false;
                }
                return form.onSubmit(value);
            }}
        >
            <div className="nice-header-input-wrapper">
                <input type="text" placeholder={ form.placeholder || null } className="nice-form-control" />
                <button type="submit">
                    <i className="fa fa-search" />
                </button>
            </div>
        </form>);
    }

    renderProduct() {
        const { product } = this.props;

        if (!product) {
            return null;
        }

        return (<span className="nice-header-product">
            { product }
        </span>);
    }

    renderMenuToggle() {
        const { responsive, collapsed } = this.props;

        if (!responsive) {
            return null;
        }

        return (<button type="button" className={ classNames('nice-header-toggle', {
            collapsed
        }) }>
            <span className="icon-bar top-bar" />
            <span className="icon-bar middle-bar" />
            <span className="icon-bar bottom-bar" />
        </button>);
    }

    render() {
        const { children, fixed, fluid } = this.props;

        return (
            <header className={ classNames('nice-header', {
                fixed
            }) }>
                <div className={ classNames('nice-header-container', {
                    fluid
                }) }>
                    { this.renderBrand() }
                    <div className="nice-header-collapse">
                        { this.renderProduct() }
                        { this.renderForm() }
                        { children }
                    </div>
                    { this.renderMenuToggle() }
                </div>
            </header>
        );
    }
}

Header.Menu = Menu;

Header.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.array
    ]),
    collapsed: PropTypes.bool,
    logo: PropTypes.shape({
        img: PropTypes.string.isRequired,
        svg: PropTypes.string,
        href: PropTypes.string,
        title: PropTypes.string,
        width: PropTypes.number,
        height: PropTypes.number
    }),
    fixed: PropTypes.bool,
    fluid: PropTypes.bool,
    form: PropTypes.shape({
        placeholder: PropTypes.string,
        onSubmit: PropTypes.func
    }),
    product: PropTypes.string,
    responsive: PropTypes.bool
};

Header.defaultProps = {
    children: null,
    collapsed: false,
    logo: null,
    fixed: false,
    fluid: false,
    form: null,
    product: null,
    responsive: true
};

export default Header;
