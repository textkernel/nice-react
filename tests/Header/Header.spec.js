import React from 'react';
import Header from '../../lib/Header';

const product = 'Nice! React Components';
const svg = '//nice.textkernel.nl/assets/textkernel.svg';
const img = '//nice.textkernel.nl/assets/textkernel@2x.png';
const width = 170;
const height = 30;
const title = 'Nice! React Components';
const href = '/';
const fixed = true;
const fluid = true;
const placeholder = 'Search...';
const responsive = true;
const collapsed = true;

describe('<Header />', () => {
    it('should render default Header', () => {
        const wrapper = shallow(<Header logo={{
            svg,
            img,
            href,
            width,
            height,
            title
        }}
        form={{
            placeholder,
            onSubmit: value => {
                window.alert(value);
                return false;
            }
        }}
        collapsed={ collapsed }
        responsive={ responsive }
        fixed={ fixed }
        fluid={ fluid }
        product={ product }
    >
        <Header.Menu>
            <Header.Menu.Item label="Menu 1" context="primary" width="125">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
                <Header.Menu.Item label="Item 3">
                    <Header.Menu.Item label="Item 3.1" />
                    <Header.Menu.Item label="Item 3.2" />
                    <Header.Menu.Item label="Item 3.3" />
                    <Header.Menu.Item label="Item 3.4" />
                </Header.Menu.Item>
                <Header.Menu.Item label="Item 4" />
            </Header.Menu.Item>
            <Header.Menu.Item label="Menu 2" context="good" width="125">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
                <Header.Menu.Item label="Item 3">
                    <Header.Menu.Item label="Item 3.1" />
                    <Header.Menu.Item label="Item 3.2" />
                    <Header.Menu.Item label="Item 3.3" />
                    <Header.Menu.Item label="Item 3.4" />
                </Header.Menu.Item>
                <Header.Menu.Item label="Item 4" />
            </Header.Menu.Item>
            <Header.Menu.Item label="Textkernel" href="https://textkernel.com" target="_blank" context="brand" width="125" className="product-icon textkernel" />
        </Header.Menu>
        <Header.Menu alternate>
            <Header.Menu.Item label="Alternate 1">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
            </Header.Menu.Item>
            <Header.Menu.Item label="Alternate 2">
                <Header.Menu.Item label="Item 1" />
                <Header.Menu.Item label="Item 2" />
            </Header.Menu.Item>
        </Header.Menu>
    </Header>);
        expect(wrapper).toMatchSnapshot();
    });
});
