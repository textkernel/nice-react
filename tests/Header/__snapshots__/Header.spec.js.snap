// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render default Header 1`] = `
<header
  className="nice-header fixed"
>
  <div
    className="nice-header-container fluid"
  >
    <h1
      className="nice-header-brand"
    >
      <a
        href="/"
        title="Nice! React Components"
      >
        <svg
          height={30}
          width={170}
        >
          <image
            height={30}
            src="//nice.textkernel.nl/assets/textkernel@2x.png"
            width={170}
            xlinkHref="//nice.textkernel.nl/assets/textkernel.svg"
          />
        </svg>
      </a>
    </h1>
    <div
      className="nice-header-collapse"
    >
      <span
        className="nice-header-product"
      >
        Nice! React Components
      </span>
      <form
        className="nice-header-form"
        onSubmit={[Function]}
      >
        <div
          className="nice-header-input-wrapper"
        >
          <input
            className="nice-form-control"
            placeholder="Search..."
            type="text"
          />
          <button
            type="submit"
          >
            <i
              className="fa fa-search"
            />
          </button>
        </div>
      </form>
      <Menu
        alternate={false}
        level={0}
      >
        <Item
          className={null}
          context="primary"
          href="#"
          label="Menu 1"
          level={0}
          onClick={null}
          target={null}
          width="125"
        >
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 1"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 2"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 3"
            level={0}
            onClick={null}
            target={null}
            width={null}
          >
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.1"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.2"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.3"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.4"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
          </Item>
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 4"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
        </Item>
        <Item
          className={null}
          context="good"
          href="#"
          label="Menu 2"
          level={0}
          onClick={null}
          target={null}
          width="125"
        >
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 1"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 2"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 3"
            level={0}
            onClick={null}
            target={null}
            width={null}
          >
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.1"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.2"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.3"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
            <Item
              className={null}
              context={null}
              href="#"
              label="Item 3.4"
              level={0}
              onClick={null}
              target={null}
              width={null}
            />
          </Item>
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 4"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
        </Item>
        <Item
          className="product-icon textkernel"
          context="brand"
          href="https://textkernel.com"
          label="Textkernel"
          level={0}
          onClick={null}
          target="_blank"
          width="125"
        />
      </Menu>
      <Menu
        alternate={true}
        level={0}
      >
        <Item
          className={null}
          context={null}
          href="#"
          label="Alternate 1"
          level={0}
          onClick={null}
          target={null}
          width={null}
        >
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 1"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 2"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
        </Item>
        <Item
          className={null}
          context={null}
          href="#"
          label="Alternate 2"
          level={0}
          onClick={null}
          target={null}
          width={null}
        >
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 1"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
          <Item
            className={null}
            context={null}
            href="#"
            label="Item 2"
            level={0}
            onClick={null}
            target={null}
            width={null}
          />
        </Item>
      </Menu>
    </div>
    <button
      className="nice-header-toggle collapsed"
      type="button"
    >
      <span
        className="icon-bar top-bar"
      />
      <span
        className="icon-bar middle-bar"
      />
      <span
        className="icon-bar bottom-bar"
      />
    </button>
  </div>
</header>
`;
